File: backend/nodejs/apps/src/libs/services/telemetry/modules/http-metrics.ts
import { metricsBackend } from '../metrics-backend';
import { SERVICE_NAME } from '../constants';

const httpRequests = metricsBackend.createCounter({
  name: 'pipeshub_http_requests_total',
  help: 'Total HTTP requests handled by the service',
  labelNames: ['service', 'route', 'method', 'status', 'org', 'domain'],
});

const httpRequestDuration = metricsBackend.createHistogram({
  name: 'pipeshub_http_request_duration_seconds',
  help: 'HTTP request latency in seconds',
  labelNames: ['service', 'route', 'method'],
  buckets: [0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 30],
});

export function recordHttpRequest(
  route: string,
  method: string,
  statusCode: number,
  orgId?: string,
  durationSeconds?: number,
  domain?: string,
): void {
  const labels = {
    service: SERVICE_NAME,
    route: route === '' ? 'unmatched' : route,
    method,
    status: String(statusCode),
    org: orgId == null || orgId === '' ? 'unknown' : orgId,
    domain: domain == null || domain === '' ? 'unknown' : domain,
  };
  httpRequests.inc(labels);
  if (typeof durationSeconds === 'number') {
    httpRequestDuration.observe(
      { service: SERVICE_NAME, route: labels.route, method },
      durationSeconds,
    );
  }
}

File: backend/python/app/telemetry/modules/http_metrics.py
"""HTTP request metrics, emitted automatically by MetricsMiddleware for every
request across every service. ``route`` is always a route template
(``/api/v1/agent/{id}``), never a raw path, to keep label cardinality bounded.
"""

from app.telemetry.backend import METRICS_BACKEND

HTTP_REQUESTS = METRICS_BACKEND.counter(
    "pipeshub_http_requests_total",
    "Total HTTP requests handled by the service",
    ["service", "route", "method", "status", "org", "domain"],
)

# Buckets span fast cache hits to slow LLM/RAG calls.
HTTP_REQUEST_DURATION = METRICS_BACKEND.histogram(
    "pipeshub_http_request_duration_seconds",
    "HTTP request latency in seconds",
    ["service", "route", "method"],
    buckets=(0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0, 30.0),
)

# Unbounded route values grow the push payload until the collector rejects it.
_MAX_UNIQUE_ROUTES = 300
_seen_routes: set[str] = set()


def bound_route(route: str) -> str:
    if route in _seen_routes:
        return route
    if len(_seen_routes) >= _MAX_UNIQUE_ROUTES:
        return "other"
    _seen_routes.add(route)
    return route
